<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>University of Sheffield PopGen Workshop March 2016 by padraicc</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">University of Sheffield PopGen Workshop March 2016</h1>
      <h2 class="project-tagline"></h2>
      <a href="https://github.com/padraicc/uspopgen" class="btn">View on GitHub</a>
      <a href="https://github.com/padraicc/uspopgen/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/padraicc/uspopgen/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="snp-calling-and-filtering" class="anchor" href="#snp-calling-and-filtering" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SNP calling and Filtering</h1>

<p>We will use some whole genome data from 10 great tit idnviduals to look at SNP calling and filtering.
These 10 individuals were sampled in Europe and area subset of the 29 birds that were sequenced and 
analysed in <a href="http://www.nature.com/ncomms/2016/160125/ncomms10474/full/ncomms10474.html">Laine et al. (2016)</a>.</p>

<p>We will focus on a small subset of the genome, calling SNPs on chrLGE22</p>

<p>The 10 birds were sequenced with 100bp paired-end Illumina reads to a mean depth of between 9-14x.
In the Laine et al. (2016) paper the SNP calling was performed using GATK, Platypus and ANGSD.</p>

<h2>
<a id="programs-required" class="anchor" href="#programs-required" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Programs required</h2>

<p>All the programs we will use are available on iceberg through the module system or from the genomics repository.</p>

<p>The following programs will be used:</p>

<ul>
<li>samtools v1.2 (available through the module system)</li>
<li>bcftools v1.3 (available in the genomics repository)</li>
<li>GATK v3.4 (available through the  module system)</li>
<li>bedtools v2.5 (available through the module system)</li>
</ul>

<h2>
<a id="obtaining-the-tutorial-material" class="anchor" href="#obtaining-the-tutorial-material" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Obtaining the tutorial material</h2>

<p>Type the following at the command terminal.</p>

<pre><code>git clone https://github.com/padraicc/uspopgen
</code></pre>

<p>This will download a folder called uspopgen into your current directory. Change directory into the uspopgen folder.</p>

<pre><code>cd uspopgen
</code></pre>

<p>Now take a look at the contents.</p>

<pre><code>ls
</code></pre>

<p>You will find folders called <code>data/</code> and <code>scripts/</code> and a <code>README.md</code> file containing the the text for this webpage.</p>

<h2>
<a id="data-files" class="anchor" href="#data-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Data files</h2>

<p>The data files we will use are located in the data folder</p>

<pre><code>ls data
</code></pre>

<p><strong>gvcf_files/</strong> contains the 10 gVCF files needed for the GATK SNP calling.<br>
<strong>bam_files/</strong> cotains the BAM files needed for the samtools SNP calling.<br>
<strong>ref_files/</strong> contains the great tit reference genome files for chrLGE22.  </p>

<h2>
<a id="snp-callers" class="anchor" href="#snp-callers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SNP callers</h2>

<p>We will call SNPs using two popular SNP calling programs: GATK and samtools/bcftools.</p>

<h3>
<a id="gatk-snp-calling" class="anchor" href="#gatk-snp-calling" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>GATK SNP calling</h3>

<p>We will perform SNP calling of the 10 birds using the shell script called <code>gatk_snp_calling.sh</code>. Due to time constraints
we will only run the last step of the GATK pipeline. Here we use the <em>GenotypeGVCFs</em> tool to produce the VCF files from the
10 gVCF files in <code>data/gvcf_files</code>. (The gVCF used as input files were produced using the GATK <em>HaplotypeCaller</em> tool.) </p>

<p>If you wish to take a look at the gatk command inside the shell script you can use <code>less</code> command as follows</p>

<pre><code>less -S scripts/gatk_snp_calling.sh
</code></pre>

<p>To run the script in the terminal just type the following:</p>

<pre><code>bash scripts/gatk_snp_calling.sh 
</code></pre>

<p>This will result in a compressed VCF file being written in the new <em>vcf_files</em> folder found in the current working 
directory. </p>

<pre><code>ls vcf_files
</code></pre>

<h3>
<a id="samtools-and-bcftools-snp-calling" class="anchor" href="#samtools-and-bcftools-snp-calling" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SAMtools and bcftools SNP calling</h3>

<p>We will also use the samtools and bcftools programs to call SNPs from the BAM file for our 10 birds. The BAM contain the
alignments of the reads mapped to the great tit chrLGE22 reference genome. The bam files are located in the <code>data/bam_files/</code>
folder. </p>

<p>To run the samtools snp calling just type the following:</p>

<pre><code>bash scripts/samtools_snp_calling.sh
</code></pre>

<h3>
<a id="understanding-the-vcf-format" class="anchor" href="#understanding-the-vcf-format" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Understanding the VCF format</h3>

<p>Now that we have run the SNP calling programs we will take a look at the VCF format files contained in the <code>vcf_files/</code>
folder.</p>

<p>The VCF format is composed of a header section where each line begins with '##' and the headers describing the columns
are located on the line starting with '#CHROM'. </p>

<p><strong>Table 1</strong> Description of VCF fields</p>

<table>
<thead>
<tr>
<th align="left">Column Number</th>
<th align="left">Title</th>
<th align="left">What it contains</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">1</td>
<td align="left">CHROM</td>
<td align="left">Chromosome/scaffold of the reference genome</td>
</tr>
<tr>
<td align="left">2</td>
<td align="left">POS</td>
<td align="left">Position on the chromosome/scaffold given in CHROM (1-based)</td>
</tr>
<tr>
<td align="left">3</td>
<td align="left">ID</td>
<td align="left">ID for the SNP/INDEL (Here always '.')</td>
</tr>
<tr>
<td align="left">4</td>
<td align="left">REF</td>
<td align="left">Allele in the reference genome</td>
</tr>
<tr>
<td align="left">5</td>
<td align="left">ALT</td>
<td align="left">The variant allele</td>
</tr>
<tr>
<td align="left">6</td>
<td align="left">QUAL</td>
<td align="left">The variant quality score (phred-scale)</td>
</tr>
<tr>
<td align="left">7</td>
<td align="left">FILTER</td>
<td align="left">THe field storing filtering information as a string ('.' in an unfiltered file)</td>
</tr>
<tr>
<td align="left">8</td>
<td align="left">INFO</td>
<td align="left">';' delimited list of variant annotations</td>
</tr>
<tr>
<td align="left">9</td>
<td align="left">FORMAT</td>
<td align="left">Describes the format for the sample genotype information in column 10 onward</td>
</tr>
<tr>
<td align="left">10 to end</td>
<td align="left">Sample id (from SM tag in BAM file RG string)</td>
<td align="left">Sample genotype information (Usually also stores the genotype qualities (GQ) and genotype likelihoods (GL or PL (phred-scaled)</td>
</tr>
</tbody>
</table>

<p>The INFO field contains a lot of annotations for the variant site (e.g Depth, Mapping Qualtity etc.) and
these values may be used for filtering (see below). The format for the genotype information is explained in the header
of VCF file. Each row following the header section is a variant site in the VCF, either a SNP or an INDEL. For more
information on the VCF format see <a href="http://samtools.github.io/hts-specs/VCFv4.2.pdf">here.</a> and <a href="http://gatkforums.broadinstitute.org/gatk/discussion/1268/what-is-a-vcf-and-how-should-i-interpret-it">here.</a></p>

<p>Now, take a look at the GATK and samtools vcf files.</p>

<pre><code>less -S vcf_files/gatk.chrLGE22.raw.snps.indels.vcf.gz
less -S vcf_files/samtools.chrLGE22.raw.snps.indels.vcf.gz 
</code></pre>

<p>If you want to exclude the header section, try the following.</p>

<pre><code>zgrep -v ^## vcf_files/gatk.chrLGE22.raw.snps.indels.vcf.gz | less -S
</code></pre>

<p>To find an explanation of the info field in the samtools VCF, look at the last 25 lines of the header section</p>

<pre><code>zgrep ^# vcf_files/samtools.chrLGE22.raw.snps.indels.vcf.gz | tail -n 25
</code></pre>

<p>Note that different SNP callers will produce some different annotations in the INFO field and genotype
fields. </p>

<p><strong>Q1.</strong> What differences do you see between the INFO field of the samtools VCF and the GATK VCF files?</p>

<p>Samtools is less well documented with regard to exactly what some of its annotations are, beyond what can be understood from the VCF header. 
GATK has a much richer documentations which can be browsed <a href="https://www.broadinstitute.org/gatk/guide/tooldocs/#VariantAnnotations">here.</a></p>

<h3>
<a id="comparing-the-output-from-the-two-callers" class="anchor" href="#comparing-the-output-from-the-two-callers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Comparing the output from the two callers</h3>

<p>We will count and compare the unfiltered SNP calls using some common command line tools and the bedtools program.</p>

<p>Count the number of variants in the vcf file using zgrep (grep for compressed files).</p>

<pre><code>zgrep -cv ^# vcf_files/gatk.chrLGE22.raw.snps.indels.vcf.gz 
zgrep -cv ^# vcf_files/samtools.chrLGE22.raw.snps.indels.vcf.gz 
</code></pre>

<p><strong>Q2.</strong> Which tool calls the most variants?</p>

<p>Now we will extract only biallelic SNPs and discard the INDELs from the VCF using GATK's SelectVariants tool.</p>

<pre><code>module load apps/binapps/GATK
java  -Xmx3g -jar $GATKHOME/GenomeAnalysisTK.jar -T SelectVariants -R data/ref_files/Parus_major_1.04.chrLGE22.fa -V vcf_files/gatk.chrLGE22.raw.snps.indels.vcf.gz -o vcf_files/gatk.chrLGE22.raw.snps.vcf.gz -selectType SNP -restrictAllelesTo BIALLELIC 
</code></pre>

<p>We will also extract the SNPs from the samtools vcf, but first we need to index it, as the GATK tools require this.</p>

<pre><code>tabix -p vcf vcf_files/samtools.chrLGE22.raw.snps.indels.vcf.gz
java  -Xmx3g -jar $GATKHOME/GenomeAnalysisTK.jar -T SelectVariants -R data/ref_files/Parus_major_1.04.chrLGE22.fa -V vcf_files/samtools.chrLGE22.raw.snps.indels.vcf.gz -o vcf_files/samtools.chrLGE22.raw.snps.vcf.gz -selectType SNP -restrictAllelesTo BIALLELIC 
</code></pre>

<p><strong>Q3.</strong> How many SNPs in the VCF files?</p>

<p>Extract and compare the number of SNPs called by each caller using bedtools. Bedtools is a very useful tools for working
with genome interval data in bed, VCF or GFF format.</p>

<pre><code>module load apps/gcc/5.2/bedtools
</code></pre>

<p>We will first extract the SNPs that were called by both callers.</p>

<pre><code>bedtools intersect -header -a vcf_files/gatk.chrLGE22.raw.snps.vcf.gz -b vcf_files/samtools.chrLGE22.raw.snps.vcf.gz | bgzip &gt; vcf_files/both.chrLGE22.raw.snps.vcf.gz
</code></pre>

<p>Next we will find SNPs called by only one of the callers.    </p>

<pre><code>bedtools subtract -header -a vcf_files/gatk.chrLGE22.raw.snps.vcf.gz -b vcf_files/samtools.chrLGE22.raw.snps.vcf.gz | bgzip &gt; vcf_files/gatk_only.chrLGE22.raw.snps.vcf.gz
bedtools subtract -header -a vcf_files/samtools.chrLGE22.raw.snps.vcf.gz -b vcf_files/gatk.chrLGE22.raw.snps.vcf.gz | bgzip &gt;  vcf_files/samtools_only.chrLGE22.raw.snps.vcf.gz
</code></pre>

<p><strong>Q4.</strong> How many SNPs did both callers call? How many by only one of the callers?</p>

<h2>
<a id="filtering-snps" class="anchor" href="#filtering-snps" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Filtering SNPs</h2>

<p>We now have the raw SNPs in VCF format. However, these files are likely to contain a number of false positives, so we need
to apply some hard filters to remove sites that may not be true SNPs. Hard filtering refers to setting a threshold on
a number of SNP properties (many of those in the listed INFO field of the VCF) and removing any SNPs that fail to meet 
that threshold.</p>

<p>Some common filters applied to SNP calls include:</p>

<ul>
<li>Depth filters (Setting a minimum and maximum depth for a site)<br>
</li>
<li>Minimum SNP quality (Requiring a minimum quality in the QUAL field of the SNP)<br>
</li>
<li>Minimum RMS mapping quality for SNPs<br>
</li>
<li>Strand Bias (Filtering sites where the number of reference and non-reference reads are highly correlated with the strands of the reads)<br>
</li>
<li>Hardy-Weinberg Equilibrium (Filter sites that show significant deviation from Hardy-Weinberg Expectations) </li>
</ul>

<p>For a summary of SNP filtering applied to whole genome resequencing studies in birds see <a href="https://www.dropbox.com/s/xa0bndtz42i1uft/snp_filtering_avian_studies.pdf?dl=0">here.</a></p>

<p>One of the simplest thresholds that can be applied is a minimum quality score. Here we will use the a <em>bcftools filter</em>
to filter our samtools SNP VCF. Documentation on this tool can be found <a href="https://samtools.github.io/bcftools/bcftools.html#filter">here.</a></p>

<pre><code>export PATH=/usr/local/extras/Genomics/apps/bcftools/1.3/bin/:$PATH
bcftools filter -e "QUAL&lt;30" -s "LOW_QUAL" vcf_files/samtools.chrLGE22.raw.snps.vcf.gz -O z -o vcf_files/samtools.chrLGE22.qual_filtered.snps.vcf.gz
</code></pre>

<p><strong>Q5.</strong> How many SNPs PASS at the QUAL &lt; 30 filter for samtools VCF?</p>

<p>(<em>Note</em> that GenotypeGVCF by default excludes calls with QUAL less than 30. See the documentation on GenotypeGVCF tools 
<a href="https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_variantutils_GenotypeGVCFs.php">here</a></p>

<p>We can also apply a depth filters with <code>bcftools filter</code>. (Note that the '||' stand for the logical 'or'). The mean depth
for sites was 106 across samples. Here we apply a filter on sites with greater than twice the mean or less than half the mean depth.</p>

<pre><code>bcftools filter -e "QUAL&lt;30 || MAX(DP)&gt;213 || MIN(DP)&lt;53 " -s "LOW_QUAL" vcf_files/samtools.chrLGE22.raw.snps.vcf.gz -O z -o vcf_files/samtools.chrLGE22.qual_dp_filtered.snps.vcf.gz
</code></pre>

<p>Count the number of filtered sites as follows.</p>

<pre><code>zgrep -v ^# vcf_files/samtools.chrLGE22.qual_dp_filtered.snps.vcf.gz | grep -cw LOW_QUAL
</code></pre>

<p><strong>Q6.</strong> How many SNPs are excluded when we apply both the depth and quality filter to the samtools VCF?   </p>

<h3>
<a id="gatk-recommended-hard-filters" class="anchor" href="#gatk-recommended-hard-filters" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>GATK recommended Hard Filters</h3>

<p>In GATK there are two options for filtering SNPs, either hard or soft filtering. The first is to use a 'soft' filtering 
approach by using known variants to carry out Variant Quality Score Recalibration. This approach requires a large number 
(100,000s) of known SNPs and may not be useful for researchers working on non-model organisms, or species without 
a large existing set of polymorphism data. Also, this approach may not be used for targeted resequencing such 
GBS or RADseq data.</p>

<p>In cases where VQSR can not be used, the GATK developers recommend a set of hard filters for filtering SNPs. Further details
on these recommended filters can be seen <a href="http://gatkforums.broadinstitute.org/gatk/discussion/2806/howto-apply-hard-filters-to-a-call-set">here</a></p>

<p>To apply the hard filters run the following command on the GATK VCF containing only SNPs.</p>

<pre><code>java -Xmx3g -jar $GATKHOME/GenomeAnalysisTK.jar -T VariantFiltration -R data/ref_files/Parus_major_1.04.chrLGE22.fa -V vcf_files/gatk.chrLGE22.raw.snps.vcf.gz --filterExpression "QD&lt;2.0||FS&gt;60.0||MQ&lt;40.0||MQRankSum&lt;-12.5||ReadPosRankSum&lt;-8.0" --filterName "GATK_hard_snp_filter" -o vcf_files/gatk.chrLGE22.hard_filtered.snps.vcf.gz
</code></pre>

<p>(<strong>Note</strong> that the VariantFiltration tool issues a warning at sites that lack sites lacking the MQRankSum and ReadPosRankSum annotations.
See why this is the case <a href="http://gatkforums.broadinstitute.org/gatk/discussion/2806/howto-apply-hard-filters-to-a-call-set">here</a>)</p>

<p>Count the number of SNPs that PASS the filters.</p>

<pre><code>zgrep -v ^# vcf_files/gatk.chrLGE22.hard_filtered.snps.vcf.gz | grep -cw PASS
</code></pre>

<p><strong>Q7.</strong> How many SNPs were filtered using the GATK recommended filters? Do the recommended GATK hardfilters filter based on th QUAL score?</p>

<h2>
<a id="region-filters" class="anchor" href="#region-filters" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Region Filters</h2>

<p>Finally, if you have time you can look at how to filter SNPs that fall in repetitive regions of the genome.</p>

<p>Another common additional filter applied to filtering SNPs in whole genome data is to exclude SNPs that fall in 
repetitive regions of the genomes where mapping may be problematic. Here we will further filter the GATK hard filtered 
VCF file to exclude SNPs in repetitive regions using a bed file located at <code>data/ref_files/chrLGE22.reps.bed</code> which 
specifies the repetitive regions of chrLGE22.</p>

<p><strong>Q8.</strong> How might you exclude SNPs in our filtered GATK VCF file within repetitive regions using bedtools?</p>

<p><em>Solutions to all the questions are available from here</em> </p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/padraicc/uspopgen">University of Sheffield PopGen Workshop March 2016</a> is maintained by <a href="https://github.com/padraicc">padraicc</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
