<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>University of Sheffield PopGen Workshop March 2016 by padraicc</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">University of Sheffield PopGen Workshop March 2016</h1>
      <h2 class="project-tagline"></h2>
      <a href="https://github.com/padraicc/uspopgen" class="btn">View on GitHub</a>
      <a href="https://github.com/padraicc/uspopgen/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/padraicc/uspopgen/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="snp-calling-and-filtering" class="anchor" href="#snp-calling-and-filtering" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SNP calling and Filtering</h1>

<p>We will use some whole genome data from 10 great tit idnviduals to look at SNP calling and filtering.
These 10 individuals were sampled in Europe and area subset of the 29 birds that were sequenced and 
analysed in <a href="http://www.nature.com/ncomms/2016/160125/ncomms10474/full/ncomms10474.html">Laine et al. (2016)</a>.</p>

<p>We will foucs on a small subset of the genome, calling SNPs on chrLGE22</p>

<p>The 10 birds were sequenced with 100bp paired-end Illumina reads to a mean depth of between 9-14x.
In the Laine et al. (2016) paper the SNP calling was performed using GATK, Platypus and ANGSD.</p>

<h2>
<a id="programs-required" class="anchor" href="#programs-required" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Programs required</h2>

<p>All the programs we will use are available on iceberg through the module system or from the genomics repository.</p>

<p>The following programs will be used:</p>

<ul>
<li>samtools v1.2 (available through the module system)</li>
<li>bcftools v1.3 (available in the genomics repository)</li>
<li>GATK v3.4 (available through the  module system)</li>
<li>ANGSD v0.911</li>
</ul>

<h2>
<a id="obtaining-the-tutorial-material" class="anchor" href="#obtaining-the-tutorial-material" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Obtaining the tutorial material</h2>

<p>Type the following at the command terminal</p>

<pre><code>git clone https://github.com/padraicc/uspopgen
</code></pre>

<p>Change directory into the uspopgen folder</p>

<pre><code>cd uspopgen
ls
</code></pre>

<p>You will find folders called data/ and scripts/ and a README.md files containing the the text for this webpage.</p>

<h2>
<a id="data-files" class="anchor" href="#data-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Data files</h2>

<p>The data files we will use are located in the data folder</p>

<pre><code>ls data
</code></pre>

<p>gvcf_files/ contains the 10 gVCF files needed for the GATK SNP calling
bam_files/ cotains the BAM files for samtools SNP calling.
ref_files/ contains the reference genome file for chrLGE22</p>

<h2>
<a id="snp-callers" class="anchor" href="#snp-callers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SNP callers</h2>

<p>We will call SNPs using GATK and samtools/bcftools.</p>

<h3>
<a id="gatk-snp-calling" class="anchor" href="#gatk-snp-calling" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>GATK SNP calling</h3>

<p>We will perform SNP calling of the 10 birds using the shell script called <strong>gatk_snp_calling.sh</strong>. Due to time constraints
we will run the last strep of the GATK pipeline. Here we use the <em>GenotypeGVCFs</em> tool to produce the VCF files from the
10 gVCF files in <em>/data/gvcf_files</em>. (The gVCF used as input files were produced using the GATK HaplotypeCaller program.) </p>

<p>To run the script in the terminal just type the following:</p>

<pre><code>bash scripts/gatk_snp_calling.sh 
</code></pre>

<p>This will result in a compressed VCF file being written in the new <em>vcf_files</em> folder found in the current working 
directory.</p>

<pre><code>ls vcf_files
</code></pre>

<h3>
<a id="samtoolsbcftools-snp-calling" class="anchor" href="#samtoolsbcftools-snp-calling" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SAMTOOLS/BCFTOOLS SNP calling</h3>

<pre><code>bash scripts/samtools_snp_calling.sh
</code></pre>

<h3>
<a id="understanding-the-vcf-format" class="anchor" href="#understanding-the-vcf-format" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Understanding the VCF format</h3>

<p>First, we will take a look at the VCF format, as this is the output format for most SNP calling and genotyping programs</p>

<p>The VCF format is composed of a header section where each line begins with '##' and the headers describing the columns
are located on the line starting with '#chrom'. The columns of the VCF from left to right are the chromosome/scaffold,<br>
position within the reference chromosome/scaffold, an ID (here always ‘.’), the reference allele, the variant allele,
the SNP/INDEL quality score, the filter field, info field, the sample format field and the samples make up the remaining
columns (10 in this case).</p>

<p>The INFO field contains a lot of annotations for the variant site (e.g Depth, Mapping Qualtity etc.) and
these values may be used for filtering (see below). The format for the genotype information is explained in the header
of VCF file. Each row following the header section is a variant site in the VCF, either a SNP or an indel.</p>

<p>Now, take a look at the GATK and samtools vcf files.</p>

<pre><code>less -S vcf_files/gatk.chrLGE22.raw.snps.indels.vcf.gz
less -S vcf_files/samtools.chrLGE22.raw.snps.indels.vcf.gz 
</code></pre>

<p>If you want to exclude the header section, try the following.</p>

<pre><code>zgrep -v ^## vcf_files/gatk.chrLGE22.raw.snps.indels.vcf.gz | less -S
</code></pre>

<p>Note that different SNP callers will have some differences in the annotations present in the INFO field and differences
in the format fields. What differences do you see between the INFO field of the samtools VCF and the GATK VCF files?</p>

<h3>
<a id="comparing-the-output-from-the-two-callers" class="anchor" href="#comparing-the-output-from-the-two-callers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Comparing the output from the two callers</h3>

<p>We will count and compare the unfiltered SNP calls using some common command line tools and the bedtools program.</p>

<p>Count the number of variants in the vcf file using zgrep (grep for compressed files).</p>

<pre><code>zgrep -cv ^# vcf_files/gatk.chrLGE22.raw.snps.indels.vcf.gz 
zgrep -cv ^# vcf_files/samtools.chrLGE22.raw.snps.indels.vcf.gz 
</code></pre>

<p>Which tool calls the most variants?</p>

<p>Now we will extract only biallelic SNPs and discard the INDELs from the VCF using GATK's SelectVariants tool</p>

<pre><code>module load apps/binapps/GATK
java  -Xmx3g -jar $GATKHOME/GenomeAnalysisTK.jar -T SelectVariants -R data/ref_files/Parus_major_1.04.chrLGE22.fa -V vcf_files/gatk.chrLGE22.raw.snps.indels.vcf.gz -o vcf_files/gatk.chrLGE22.raw.snps.vcf.gz -selectType SNP -restrictAllelesTo BIALLELIC 
</code></pre>

<p>We will extract the SNPs from the samtools vcf, but first we need to index it, as the GATK tools require this.</p>

<pre><code>tabix -p vcf vcf_files/samtools.chrLGE22.raw.snps.indels.vcf.gz
java  -Xmx3g -jar $GATKHOME/GenomeAnalysisTK.jar -T SelectVariants -R data/ref_files/Parus_major_1.04.chrLGE22.fa -V vcf_files/samtools.chrLGE22.raw.snps.indels.vcf.gz -o vcf_files/samtools.chrLGE22.raw.snps.vcf.gz -selectType SNP -restrictAllelesTo BIALLELIC 
</code></pre>

<p>How many SNPs in the VCF files?</p>

<p>Extract and compare the number of SNPs called by each caller using bedtools. Bedtools is a very useful tools for working
with genome interval data in bed, VCF or GFF format.</p>

<pre><code>module load apps/gcc/5.2/bedtools
bedtools intersect -header -a vcf_files/gatk.chrLGE22.raw.snps.vcf.gz -b vcf_files/samtools.chrLGE22.raw.snps.vcf.gz &gt; vcf_files/both.chrLGE22.raw.snps.vcf.gz
</code></pre>

<h2>
<a id="filtering-snps" class="anchor" href="#filtering-snps" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Filtering SNPs</h2>

<h3>
<a id="depth-filters" class="anchor" href="#depth-filters" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Depth filters</h3>

<h3>
<a id="base-qualities-strand-biases" class="anchor" href="#base-qualities-strand-biases" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Base qualities, Strand Biases</h3>

<h3>
<a id="region-filters" class="anchor" href="#region-filters" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Region Filters</h3>

<h3>
<a id="gatk-recommended-hard-filters" class="anchor" href="#gatk-recommended-hard-filters" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>GATK recommended Hard Filters</h3>

<h2>
<a id="angsd-snp-calling-and-popgen-analysis-in-low-coverage-data" class="anchor" href="#angsd-snp-calling-and-popgen-analysis-in-low-coverage-data" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>ANGSD (SNP calling and Popgen analysis in low coverage data)</h2>

<h2>
<a id="sfs" class="anchor" href="#sfs" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SFS</h2>

<p>Comparing sfs</p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/padraicc/uspopgen">University of Sheffield PopGen Workshop March 2016</a> is maintained by <a href="https://github.com/padraicc">padraicc</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
